activerecord/test/cases/database_configurations_test.rb

```diff
@@ -84,6 +84,50 @@ def test_find_db_config_prioritize_db_config_object_for_the_current_env
     assert_equal ActiveRecord::ConnectionHandling::DEFAULT_ENV.call, config.env_name
     assert_equal ":memory:", config.database
   end
+
+  class CustomHashConfig < ActiveRecord::DatabaseConfigurations::HashConfig
+    def sharded?
+      custom_config.fetch("sharded", false)
+    end
+
+    private
+      def custom_config
+        configuration_hash.fetch(:custom_config)
+      end
+  end
+
+  def test_registering_a_custom_config_object
+    previous_handlers = ActiveRecord.db_config_handlers
+
+    ActiveRecord.register_db_config_handler do |env_name, name, _, config|
+      next unless config.key?(:custom_config)
+      CustomHashConfig.new(env_name, name, config)
+    end
+
+    configs = ActiveRecord::DatabaseConfigurations.new({
+      "test" => {
+        "config_1" => {
+          "database" => "db",
+          "custom_config" => {
+            "sharded" => 1
+          }
+        },
+        "config_2" => {
+          "database" => "db"
+        }
+      }
+    }).configurations
+
+    custom_config = configs.first
+    hash_config = configs.last
+
+    assert custom_config.is_a?(CustomHashConfig)
+    assert hash_config.is_a?(ActiveRecord::DatabaseConfigurations::HashConfig)
+
+    assert custom_config.sharded?
```

Comment:
I noticed that you set `sharded` to `1` in the test config, instead of true/false.

```suggestion
    assert_equal(1, custom_config.sharded?)
```

Reply:
Rails is truthy/falsey so this is fine but I realized I should use `assert_predicate` instead.